# Clase Semana 7

In [6]:
archivo = open("saludo.txt", mode="a")
archivo.write("Esto es una tercera linea\n")
archivo.close()

print("Se ha escrito en el archivo saludo.txt")

Se ha escrito en el archivo saludo.txt


In [7]:
archivo = open("notastxt", mode="w")
archivo.write("Esto es una primera linea\n")
archivo.write("Esto es una segunda linea\n")
archivo.close()

In [8]:
with open("bitacora.txt", mode="w") as archivo:
    archivo.write("Esto es una primera linea\n")
    archivo.write("Esto es una segunda linea\n")
    archivo.write("Esto es una tercera linea\n")

with open("bitacora.txt", mode="a") as archivo:
    archivo.write("Esto es una cuarta linea\n")


In [9]:
nombre = input("Ingrese su nombre")
edad = input("Ingrese su edad")

with open("datos_usuario.txt", mode="w") as archivo:
    archivo.write(f"Nombre: {nombre}, Edad: {edad}\n")

print("Datos guardados")

Datos guardados


# Lectura

In [11]:
archivo = open("Cuento.txt", mode="w")


In [1]:
with open("Cuento.txt", mode="r") as a:
    contenido = a.read()
    print(contenido) 

Habia una vez
una computadora que se apagaba 
cuando no tenia corriente electrica.
Fin.


18.1 Lectura completa con read()
El método read() lee todo el contenido del archivo y lo devuelve como una sola cadena de texto.

In [2]:
with open("Cuento.txt", "r") as archivo:
    contenido = archivo.read()
print(contenido)

Habia una vez
una computadora que se apagaba 
cuando no tenia corriente electrica.
Fin.


8.2 Lectura línea a línea con readline()
El método readline() lee una sola línea (hasta el siguiente salto de línea \n). Cada llamada avanza a la siguiente línea.

In [4]:
with open("Cuento.txt", "r") as archivo:
    linea1 = archivo.readline()
    linea2 = archivo.readline()

print("Primera linea:", linea1.strip())
print("Segunda linea:", linea2.strip())


Primera linea: Habia una vez
Segunda linea: una computadora que se apagaba


El método strip() elimina los espacios y el salto de línea (\n) que quedan al inicio y al final de cada línea leída, de modo que la impresión se vea limpia.



18.3 Lectura de todas las líneas con readlines()
El método readlines() lee todas las líneas y las devuelve en una lista, donde cada elemento es una línea (incluyendo su \n).

In [8]:
with open("Cuento.txt", "r") as archivo:
    lineas = archivo.readlines()

print("Total de lineas:", len(lineas))
for i, linea in enumerate(lineas, 1):
    print(f"{i}: {linea.strip()}")

Total de lineas: 4
1: Habia una vez
2: una computadora que se apagaba
3: cuando no tenia corriente electrica.
4: Fin.


18.4 Iterar sobre las líneas con for
La forma más eficiente de recorrer un archivo es iterar directamente sobre el objeto de archivo con un ciclo for: en cada vuelta, linea toma el valor de una línea del archivo.

In [10]:
with open("Cuento.txt", "r") as archivo:
    for linea in archivo:
        print(linea.strip())

Habia una vez
una computadora que se apagaba
cuando no tenia corriente electrica.
Fin.


# TEMA 19: INTEGRACIÓN DE LECTURA Y ESCRITURA DE ARCHIVOS

19.1 El flujo completo: leer → procesar → escribir
Los programas de aplicación reales combinan los dos temas anteriores: leen datos de un archivo, los procesan en memoria y guardan los resultados en otro archivo (o los anexan al mismo). Este flujo sigue el modelo Entrada-Proceso-Salida visto en la Semana 1:

[ Archivo de entrada ]  --->  (lectura con 'r')  --->  [ Procesamiento ]  --->  (escritura con 'w' o 'a')  --->  [ Archivo de salida ]

19.2 Ejemplo: reporte de ventas desde un archivo
Supongamos que la tienda registra sus ventas del día en ventas.txt:

Contenido de ventas.txt:

100.0
250.5
49.5

El siguiente programa lee cada monto, calcula el total y el promedio, y guarda el resumen en un nuevo archivo resumen.txt:

In [32]:
archivo = open("ventas.txt", mode="w")
archivo.write("100.0\n")
archivo.write("250.5\n")
archivo.write("49.5\n")
archivo.close()

In [33]:
# LECTURA: procesamos los montos
with open("ventas.txt", "r") as archivo:
    montos = [float(linea.strip()) for linea in archivo]

# PROCESO: calculamos total y promedio
total = sum(montos)
promedio = total / len(montos) 

# ESCRITURA: guardamos el resumen en un archivo nuevo
with open("resumen.txt", "w") as archivo:
    archivo.write(f"Total de ventas: {total:.2f}\n")
    archivo.write(f"Promedio por venta: {promedio:.2f}\n")

# VERIFICACIÓN: Leemos y mostramos el resultado guardado
with open("resumen.txt", "r") as archivo:
    print(archivo.read())

Total de ventas: 400.00
Promedio por venta: 133.33



Contenido resultante de resumen.txt:
Total de ventas: 400.00
Promedio por venta: 133.33

Observa cómo el mismo programa usa tres veces la sentencia with: una para leer, otra para escribir y otra para verificar lo escrito.

19.3 Copiar el contenido de un archivo a otro

In [16]:
archivo = open("origen.txt", mode="w")

In [17]:
with open("origen.txt", "r") as origen:
    contenido = origen.read()

with open("destino.txt", "w") as destino:
    destino.write(contenido)

print("Copia realizada.")

Copia realizada.


Contenido de origen.txt:
Contenido a copiar

Contenido resultante de destino.txt: idéntico al de origen.txt. 

Salida en consola: Copia realizada.

19.4 Anexar resultados al mismo archivo de datos
Cuando queremos conservar los datos originales y agregar un resultado, combinamos la lectura (para calcular) con la escritura en modo 'a' (para anexar):

In [34]:
with open("ventas.txt", "r") as archivo:
    montos = [float(linea.strip()) for linea in archivo]

total = sum(montos)

with open("ventas.txt", "a") as archivo:
    archivo.write(f"Total: {total:.2f}\n")

print("Total anexado al archivo ventas.txt") 

Total anexado al archivo ventas.txt


Contenido resultante de ventas.txt (después de ejecutar):
100.0
250.5
49.5
Total: 400.00


19.5 Buenas prácticas al integrar lectura y escritura
    1. Usa siempre la sentencia with para que los archivos se cierren automáticamente.
    2. Usa strip() al leer para eliminar los saltos de línea y convertir con int() o float() según el tipo de dato.
    3. Separa el código en tres fases claras: lectura, procesamiento y escritura.
    4. Si el archivo de salida no debe conservar datos previos, usa 'w'; si debe conservarlos, usa 'a'.

# TEMA 20: EXCEPCIONES EN MANEJO DE ARCHIVOS

20.1 ¿Qué es una excepción?
Una excepción es un error que ocurre durante la ejecución del programa y que, si no se controla, lo detiene de forma abrupta mostrando un traceback. Con el bloque try-except podemos "atrapar" esos errores y responder con mensajes amigables en lugar de que el programa se detenga.

20.2 Excepciones más comunes al trabajar con archivos

Excepción	            ¿Cuándo ocurre?
FileNotFoundError       Se intenta abrir un archivo que no existe o escribiren una carpeta inexistente.

PermissionError         El sistema operativo niega el acceso (archivo solo de lectura, sin permisos, o se intenta abrir un directorio como archivo).

ValueError             Al convertir el contenido leído (ej. una línea no numérica con float()).
 

20.3 Manejo básico con try-except

In [35]:
try: 
    with open("inexixtente.txt", "r") as archivo:
        print(archivo.read())
except FileNotFoundError:
    print("Error: el archivo no existe en la carpeta")

# El programa no se detiene: muestra el mensaje controlado y continúa.

Error: el archivo no existe en la carpeta


20.4 Varios except para distintos errores
Podemos controlar cada tipo de error con un bloque except específico y, al final, un except general como red de seguridad:

In [36]:
try: 
    with open("datos.txt", "r") as archivo:
        contenido = archivo.read()
    print("Lectura exitosa")
    print(contenido)
except FileNotFoundError:
    print("Error: el 'datos.txt' no existe.")
except PermissionError:
    print("Error: no tienes permisos para abrir 'datos.txt'.")
except Exception as error:
    print(f"Ocurrio un error inesperado {error}")

Error: el 'datos.txt' no existe.


20.5 La cláusula else: código que se ejecuta solo si no hubo error

In [37]:
try:
    with open("Cuento.txt", "r") as archivo:
        contenido = archivo.read()
except FileNotFoundError:
    print("El archivo no existe.")
else:
    print("El archivo se leyó correctamente.")
    print(contenido)

El archivo se leyó correctamente.
Habia una vez
una computadora que se apagaba 
cuando no tenia corriente electrica.
Fin.


20.6 Ejemplo integrador: lectura segura con nombre ingresado por el usuario

In [39]:
nombre_archivo = input("Ingresa el nombre del archivo que quieres abrir:")
try:
    with open(nombre_archivo, "r") as archivo:
        for linea in archivo:
            print(linea.strip())
except FileNotFoundError:
    print(f"Error: no se encontró el archivo '{nombre_archivo}'.")
except PermissionError:
    print(f"Error. no tienes permisos para abrir '{nombre_archivo}'.")

Habia una vez
una computadora que se apagaba
cuando no tenia corriente electrica.
Fin.


Si el usuario escribe un nombre inexistente, el programa muestra un mensaje controlado y no se detiene. Este control de excepciones es uno de los requerimientos obligatorios de la Entrega Final del Proyecto.